ovirt_imageio/client/_upload.py

```diff
@@ -0,0 +1,163 @@
+        help="Name of the storage domain.")
+
+    cmd.add_argument(
+        "-f", "--format",
+        choices=("raw", "qcow2"),
+        default="qcow2",
+        help="Upload disk format (default qcow2 for data disks and raw "
+             "for iso disks).")
+
+    cmd.add_argument(
+        "--preallocated",
+        dest="sparse",
+        action="store_false",
+        help="Create preallocated disk. Required when using raw format on "
+             "block based storage domain (iSCSI, FC). ISO images are "
+             "always uploaded to preallocated disk.")
+
+    cmd.add_argument(
+        "--disk-id",
+        type=_options.UUID,
+        help="A UUID for the new disk. If not specified oVirt will "
+             "create a new UUID.")
+
+    cmd.add_argument(
+        "filename",
+        type=_options.File,
+        help="Path to image to upload. Supported formats: raw, qcow2, iso.")
+
+
+def upload_disk(args):
+    with _ui.ProgressBar(phase="inspecting image") as progress:
+        disk_info = _prepare(args)
+        con = _ovirt.connect(args)
+        with closing(con):
+            progress.phase = "creating disk"
+            disk = _ovirt.add_disk(
+                con=con,
+                name=disk_info.name,
+                initial_size=disk_info.initial_size,
+                provisioned_size=disk_info.provisioned_size,
+                sd_name=args.storage_domain,
+                id=args.disk_id,
+                sparse=disk_info.sparse,
+                enable_backup=disk_info.format == _ovirt.COW,
+                content_type=disk_info.content_type,
+                format=disk_info.format)
+
+            progress.phase = "creating transfer"
+            host = _ovirt.find_host(con, args.storage_domain)
+            transfer = _ovirt.create_transfer(con, disk, host=host)
+            try:
+                progress.phase = "uploading image"
+                _api.upload(
+                    args.filename,
+                    transfer.transfer_url,
+                    args.cafile,
+                    buffer_size=args.buffer_size,
+                    progress=progress,
+                    proxy_url=transfer.proxy_url,
+                    max_workers=args.max_workers)
+            except Exception:
+                progress.phase = "cancelling transfer"
+                _ovirt.cancel_transfer(con, transfer)
+                raise
+
+            progress.phase = "finalizing transfer"
+            _ovirt.finalize_transfer(con, transfer, disk)
+        progress.phase = "upload completed"
+
+
+def _prepare(args):
+    # Obtain the image info dictionary.
+    img_info = _api.info(args.filename)
+    if img_info["format"] not in ("qcow2", "raw"):
+        raise RuntimeError(f"Unsupported image format {img_info['format']}")
+
+    # Obtain all remaining fields for DiskInfo.
+    initial_size = None
+    if args.format == "qcow2":
+        initial_size = _api.measure(args.filename, args.format)["required"]
```

Comment:
When uploading iso, we start with args.format == "qcow2", but at the end this function we replace the format with "raw", so we do unneeded measure and use wrong initial_size value.

So we need to check the content type and select the format before we measure.